Temporary Device Memory Type

tomstitt opened the pull request:
Adds optional device temporary memory type that can be used to prevent repeated device mallocs
 - Example in *mesh.cpp*'s `GeometricFactors::GeometricFactors`

Updates to the `Umpire{Host,Device}MemorySpace`
- Switches the pool identifier from a string to int
  - I'm okay switching back to strings but I think ints are easier to deal with
- Support setting allocator ids before init 
- Adds option to use default `HOST`/`DEVICE` `MemoryTypes` even if mfem is built with umpire
- Lazy init of `UmpireHostMemorySpace`
- Adds temporary pool identifier

I'm very happy to modify this but there is something I'd like some advice on.

For now we will need to use the default `HOST` allocator with the Umpire device allocators because using anything else causes a big performance hit. The difficult part will be with [this switch block](https://github.com/mfem/mfem/blob/master/general/mem_manager.cpp#L66-L83); we want `case MemoryType::HOST` to return `MemoryType::DEVICE_UMPIRE` if (device) umpire is enabled but this isn't very general.

Here are some runtime numbers for a 2 GPU problem of moderate size with different host allocators in mfem:

_All of these are using `DEVICE_UMPIRE`_ for the device type

| host `mfem::MemoryType` | notes                           | runtime (s)    |
|--------------------|---------------------------------|----------------|
|`HOST_UMIPRE`       | w/ umpire dynamic pools         | 239            |
|`HOST_UMPIRE`       | w/ umpire "HOST" allocator      | 211.5          |
|`HOST`              | w/ indirection*                 | 168.5          |
|`HOST`              | w/o indirection*                | 151.2          |

- By indirection I mean explicitly using `MemoryType::HOST` vs creating and using a `HostMemorySpace` object.
  - To test this I locally changed [this line](https://github.com/mfem/mfem/compare/feature/tomstitt/temp-mem-type#diff-bd2d3763de3a734165f0c0a6454ea785R681) to return a `HostMemorySpace`
  - You miss out of faster returns, [like here](https://github.com/mfem/mfem/blob/master/general/mem_manager.hpp#L686-L687) if you aren't explicitly using `MemoryType::HOST`


#### TODO
- [x] Support arbitrary host_type/device_type pairs in `MemoryType MemoryManager::GetDualMemoryType_(MemoryType mt)`
- [x] Generalize HOST_PINNED type to support hip like device types
<!--GHEX{"id":1644,"author":"tomstitt","editor":"tzanio","reviewers":["v-dobrev","camierjs","artv3"],"assignment":"2021-01-19T14:06:25-08:00","approval":"2021-03-24T22:12:03.747Z","merge":"2021-03-25T04:43:50.414Z"}XEHG--><!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1644](https://github.com/mfem/mfem/pull/1644) | @tomstitt | @tzanio | @v-dobrev + @camierjs + @artv3 | 01/19/21 | 03/24/21 | 03/24/21 | |
<!--ELBATXEHG-->

artv3 commented:
@dylan-copeland this should help with temporary memory on the gpu.

v-dobrev commented:
@tomstitt, when this is ready for review (in particular, there's nothing more you want to add), please add the "ready-for-review" label and remove "WIP".

camierjs commented:
1. Are you using `temp` other than the on the `Device` side?
I'm trying to see how a pool might be added, and was wondering if we want to support that as an `option` of some memory backends, or just as an additional one.
I tried it with a few examples (6 and the minimal surface), but haven't seen yet a huge benefit.

2. Did the `pinned` backend helps?

tomstitt commented:
>     1. Are you using `temp` other than the on the `Device` side?
>        I'm trying to see how a pool might be added, and was wondering if we want to support that as an `option` of some memory backends, or just as an additional one.
>        I tried it with a few examples (6 and the minimal surface), but haven't seen yet a huge benefit.
No we are not, only on the `Device`. The only place I think it might be useful is with a pool of pinned `Host` memory, since that is actually expensive to allocate and might be more limited. I haven't tried that yet though.

>     2. Did the `pinned` backend helps?
Yes! I've tested it in one instance where we `Read` a buffer to the device each timestep and it was ~5-6x faster, which was noticeable on the timeline given how big the object was. I'll update with some numbers when I open the results in the profiler again.

camierjs commented:
When do you need the `DeleteDevice` and `UpdateMemoryType`?

tomstitt commented:
> When do you need the `DeleteDevice` 

`DeleteDevice` is being used in the case where we have a large object that is setup once and is then used read-only on the device for a short period each timestep. We `Read` it from the host into temporary device memory, use it, then release the temporary device memory; I just called it `DeleteDevice` since that's what it does in general.

> ... and `UpdateMemoryType`?

`UpdateMemoryType`, which should probably be called `UpdateDeviceMemoryType`, is being used when it's difficult to set an object's (device) `MemoryType` initially. Specifically I'm using it when we have a `ParBilinearForm` and want its `SparseMatrix` on the device but we want it to use temporary device memory. It was easier to add this then to modify `ParBilinearForm` to support setting its `SparseMatrix`'s `MemoryType`. I can also see us wanting to change the `Host` type (to pinned memory) but that would involve a copy, which is less ideal.

camierjs commented:
I see - but we might want to be sure that no other alias is _alive_ before deleting the base: it is dangerous to remove/update any memory. It may be ok most of the time, but some function called in an `MFEM_ASSERT` that verifies consistency might be good to add if we decide to follow that path.

v-dobrev commented on general/cuda.hpp:
```diff
@@ -63,9 +63,15 @@ void* CuMemAlloc(void **d_ptr, size_t bytes);
 void* CuMallocManaged(void **d_ptr, size_t bytes);
 
+/// Allocate page-locked (pinned) host memory
+void* CuMemAllocHost(void **ptr, size_t bytes);
```
I'd call this `CuMemAllocPinned` or `CuMemAllocHostPinned`.

v-dobrev commented on general/cuda.hpp:
```diff
@@ -63,9 +63,15 @@ void* CuMemAlloc(void **d_ptr, size_t bytes);
 void* CuMemFree(void *d_ptr);
 
+/// Frees page-locked (pinned) host memory and returns destination ptr.
+void* CuMemFreeHost(void *ptr);
```
Same here, I'd call this `CuMemFreePinned` or `CuMemFreeHostPinned`.

tomstitt replied:
Done

v-dobrev commented on general/device.hpp:
```diff
@@ -262,10 +261,26 @@ class Device
    static bool GetGPUAwareMPI() { return Get().mpi_gpu_aware; }
+
+#ifdef MFEM_USE_UMPIRE
+   static bool UseHostUmpire() { return Get().use_host_umpire; }
+   static void UseHostUmpire(bool use) { Get().use_host_umpire = use; }
+   static bool UseDeviceUmpire() { return Get().use_device_umpire; }
+   static void UseDeviceUmpire(bool use) { Get().use_device_umpire = use; }
```
I personally prefer these to have `Get`/`Set` prefixes similar to other methods in this class.

v-dobrev commented on general/mem_manager.hpp:
```diff
@@ -316,6 +318,11 @@ class Memory
        i.e. it will, generally, not be Empty() after this call. */
    inline void Delete();
 
+   /** @brief Delete the owned device pointer. */
+   inline void DeleteDevice();
```
I think the description (and the behavior) should be something like this:
"Delete the device pointer, if owned. If the data is valid only on device, move it to host before deleting and invalidating the device memory."

v-dobrev replied:
Or it maybe more flexible to have an optional boolean parameter `copy_to_host` which can be `true` by default.

v-dobrev commented:
> > When do you need the `DeleteDevice`
> 
> `DeleteDevice` is being used in the case where we have a large object that is setup once and is then used read-only on the device for a short period each timestep. We `Read` it from the host into temporary device memory, use it, then release the temporary device memory; I just called it `DeleteDevice` since that's what it does in general.
> 
> > ... and `UpdateMemoryType`?
> 
> `UpdateMemoryType`, which should probably be called `UpdateDeviceMemoryType`, is being used when it's difficult to set an object's (device) `MemoryType` initially. Specifically I'm using it when we have a `ParBilinearForm` and want its `SparseMatrix` on the device but we want it to use temporary device memory. It was easier to add this then to modify `ParBilinearForm` to support setting its `SparseMatrix`'s `MemoryType`. I can also see us wanting to change the `Host` type (to pinned memory) but that would involve a copy, which is less ideal.

Regarding the use case for `UpdateMemoryType`: I think it may be better to handle this as an option for the `SparseMatrix` class -- this option will tell the class what device allocations to use, regular device or temporary device. For this reason (and in general) we should add a new `MemoryClass`, e.g. `DEVICE_TEMP` which will include just the temporary device memory types. This parameter, e.g. called `use_temp`, can then be propagated to the global functions (as a new parameter which is `false` by default):
```c++
template <typename T> inline const T *Read(const Memory<T> &mem, int size, bool on_dev = true);
template <typename T> inline T *Write(Memory<T> &mem, int size, bool on_dev = true);
template <typename T> inline T *ReadWrite(Memory<T> &mem, int size, bool on_dev = true);
```
In these functions, when `on_dev` and `use_temp` are both `true`, they will use the global setting from `Device::GetDeviceTempMemoryClass()`.

The same flag (which can be stored in the `Memory::flags` similar to the `USE_DEVICE` flag) can be used in other classes like `Array` and `Vector` in a similar way.

What do you think?

v-dobrev commented on linalg/densemat.hpp:
```diff
@@ -768,9 +775,9 @@ class DenseTensor
    int TotalSize() const { return SizeI()*SizeJ()*SizeK(); }
 
-   void SetSize(int i, int j, int k)
+   void SetSize(int i, int j, int k, MemoryType mt_ = MemoryType::SIZE)
```
I think it is better for clarity to add a new constant in `MemoryType`, say `MemoryType::PRESERVE` which can be used only in certain contexts like this.

tomstitt replied:
Yeah that makes sense, we've been talking about that in GH-1669 as well. I like `PRESERVE`, I'll add that

v-dobrev commented on mesh/mesh.cpp:
```diff
@@ -10545,17 +10545,17 @@ GeometricFactors::GeometricFactors(const Mesh *mesh, const IntegrationRule &ir,
    if (flags & GeometricFactors::COORDINATES)
    {
-      X.SetSize(vdim*NQ*NE);
+      X.SetSize(vdim*NQ*NE, Device::GetDeviceTempMemoryType());
```
The type of memory use here (regular or temp) should probably be an option in the class.

tomstitt replied:
Done!

tomstitt commented:
@v-dobrev and @camierjs, I changed the style of this a bit, there is no longer a `DEVICE_TEMP_UMPIRE` type, instead objects can be constructed with `MemoryClass::DEVICE` for device permanent memory or `MemoryClass::DEVICE_TEMP` for device temporary memory. The `DeviceMemorySpace`s have an extra `AllocTemp` method that uses `Alloc` by default but does the right thing in the `UmpireDeviceMemorySpace`.
There is also a `UseTemporary` method on `Memory` objects for setting the device temporary memory status after construction. I added an example at `examples/cuda_umpire.cpp` (I'll move to _tests_ later) that shows usage of each. Please let me know what you think! I can provide a build script on the rz if either of you want to try out the example there.

tzanio commented on examples/cuda_umpire.cpp:
```diff
@@ -0,0 +1,118 @@
+//#if defined(MFEM_USE_UMPIRE) && (defined(MFEM_USE_CUDA) || defined(MFEM_USE_HIP))
```
This is failing Travis now: https://travis-ci.com/github/mfem/mfem/jobs/471197181#L818. Should it be guarded?

Also: please add the Copyright header :)

tzanio replied:
I was going to say this is not the place for this, but I just noticed https://github.com/mfem/mfem/pull/1644#issuecomment-759148473 :)

Please move to `tests/`.

tomstitt replied:
I'll move it once the interface is set, I had some trouble running it quickly from the test system, is there a way to quickly run 1 test once if it's there?

tzanio commented on config/defaults.mk:
```diff
@@ -143,6 +143,7 @@ MFEM_USE_RAJA          = NO
 MFEM_USE_OCCA          = NO
 MFEM_USE_CEED          = NO
 MFEM_USE_UMPIRE        = NO
+MFEM_USE_CAMP          = NO
```
Is CAMP now a requirement? In which cases?

Either way, it should probably be documented in INSTALL

tomstitt replied:
It's required for recent Umpire releases, I'll add a note to INSTALL

tzanio commented on examples/makefile:
```diff
@@ -22,7 +22,7 @@ MFEM_LIB_FILE = mfem_is_not_built
 
 SEQ_EXAMPLES = ex1 ex2 ex3 ex4 ex5 ex6 ex7 ex8 ex9 ex10 ex14 ex15 ex16 ex17\
-  ex18 ex19 ex20 ex21 ex22 ex23 ex24 ex25 ex26 ex27
+  ex18 ex19 ex20 ex21 ex22 ex23 ex24 ex25 ex26 ex27 cuda_umpire
```
See above

tzanio commented on general/cuda.hpp:
```diff
@@ -63,9 +63,15 @@ void* CuMemAlloc(void **d_ptr, size_t bytes);
 /// Allocates managed device memory
 void* CuMallocManaged(void **d_ptr, size_t bytes);
 
+/// Allocate page-locked (pinned) host memory
```
For consistency

```suggestion
/// Allocates page-locked (pinned) host memory
```

tzanio commented on general/device.hpp:
```diff
@@ -123,20 +123,27 @@ class Device
+#endif
 
-   MODES mode;
+   MODES mode{Device::SEQUENTIAL};
```
Can we not mix initialization styles?

tomstitt replied:
Can you clarify: do you mean use either ` = val` or `{val}` but not both? Is one preferred over the other?

tzanio replied:
Yes the current style is `= val`.

tomstitt replied:
Updated

tzanio commented on general/mem_manager.hpp:
```diff
@@ -32,13 +32,15 @@ enum class MemoryType
    HOST_64,        ///< Host memory; aligned at 64 bytes
    HOST_DEBUG,     ///< Host memory; allocated from a "host-debug" pool
    HOST_UMPIRE,    ///< Host memory; using Umpire
+   HOST_PINNED,    ///< Host memory: pinned (page-locked)
```
These and other changes in this PR should be mentioned in `CHANGELOG`

tomstitt replied:
I added some stuff to the `CHANGELOG`, is the level of detail okay? I didn't add anything on the class constructor changes that now let you specify temporary memory but I can add that if you want.

tzanio replied:
Yes, thank you

artv3 commented on examples/cuda_umpire.cpp:
```diff
@@ -0,0 +1,118 @@
+   return a.getCurrentSize();
+}
+
+int main(int argc, char ** argv)
```
I think we need friendlier example here with more docs.

tomstitt replied:
This isn't going to stay as an example (going to be a test) but it was easier to have it here when running repeatedly (for interface updates)

tzanio commented on general/mem_manager.hpp:
```diff
@@ -316,6 +344,11 @@ class Memory
+   /** Delete the device pointer, if owned. If @a copy_to_host is true and
+    *  the data is valid only on device, move it to host before deleting.
+    *  Invalidates the device memory. */
+   inline void DeleteDevice(bool copy_to_host=true);
```
I'll drop the leading `*` on rows 2-3 and wrapm at 80 cols.

tzanio commented on makefile:
```diff
@@ -332,8 +332,8 @@ MFEM_DEFINES = MFEM_VERSION MFEM_VERSION_STRING MFEM_GIT_STRING MFEM_USE_MPI\
- MFEM_USE_MKL_CPARDISO MFEM_USE_AMGX MFEM_USE_MUMPS MFEM_SOURCE_DIR\
- MFEM_INSTALL_DIR
+ MFEM_USE_MKL_CPARDISO MFEM_USE_AMGX MFEM_USE_MUMPS MFEM_USE_CAMP \
+ MFEM_SOURCE_DIR MFEM_INSTALL_DIR
```
Is there CAMP support in the CMake build system?

tomstitt replied:
No... I'll add that

tomstitt replied:
I did some testing and we don't actually need CAMP support in CMake if Umpire is built with its CAMP submodule.  This is also true with RAJA.

If RAJA or Umpire is built with an external CAMP the build fails unless `camp_DIR` is defined.

tzanio reviewed (commented):
Please update `CHANGELOG`

tomstitt commented on general/device.cpp:
```diff
@@ -181,6 +178,8 @@ Device::~Device()
    Get().device_mem_type = MemoryType::HOST;
    Get().device_mem_class = MemoryClass::HOST;
+   Get().device_temp_mem_type = MemoryType::HOST;
+   Get().device_temp_mem_class = MemoryClass::HOST;
```
I think it would make sense to get rid of the `device_temp_mem_type` and everything related to it and just keep the `device_temp_mem_class`. The assumption with these changes is that the `DeviceMemorySpace` can optionally support temporary memory but that the temporary `MemoryType` is _not_ distinct from the permanent `MemoryType`. What do others think?

We could allow for combinations for permanent and temporary device memory types but that brings back the need for a `TEMP` `MemoryType` for each device `MemoryType` that has a temp option.

camierjs commented on INSTALL:
```diff
@@ -459,9 +458,13 @@ MFEM_USE_UMPIRE = YES/NO
    discovery, provision, and management of memory on machines with multiple
    memory devices like NUMA and GPUs.
 
+MFEM_USE_CAMP = YES/NO
```
Is there some required CAMP code needed here for MFEM to support Umpire >= 3.0.0 ?

artv3 replied:
I believe both RAJA and Umpire use CAMP in some capacity.

camierjs replied:
Sure, but that's Umpire/Raja tpl requirements that should be pulled/compiled by their build system - if we don't have any code here?

tomstitt replied:
At least with Umpire we need the CAMP include flags when building MFEM with gmake

camierjs commented on general/mem_manager.cpp:
```diff
@@ -56,6 +56,7 @@ MemoryType GetMemoryType(MemoryClass mc)
       case MemoryClass::HOST_32: return MemoryType::HOST_32;
       case MemoryClass::HOST_64: return MemoryType::HOST_64;
       case MemoryClass::DEVICE:  return mm.GetDeviceMemoryType();
+      case MemoryClass::DEVICE_TEMP:  return mm.GetDeviceTempMemoryType();
```
Is `TEMP` vs `POOL` more understandable?

tomstitt replied:
Do you mean `TEMP` or `POOL` vs `DEVICE_TEMP`? Or `DEVICE_TEMP` vs `DEVICE_POOL`?

camierjs commented on general/mem_manager.cpp:
```diff
@@ -87,6 +91,9 @@ static void MFEM_VERIFY_TYPES(const MemoryType h_mt, const MemoryType d_mt)
    MFEM_ASSERT(IsDeviceMemory(d_mt),"");
    const bool sync =
+      (h_mt == MemoryType::HOST && d_mt == MemoryType::DEVICE_UMPIRE) ||
+      (h_mt == MemoryType::HOST_PINNED && d_mt == MemoryType::DEVICE_UMPIRE) ||
```
`HOST_PINNED` type could be mixed with other `device` types too, but that can be added when we'll add the equivalent tests.

camierjs commented on general/mem_manager.hpp:
```diff
@@ -32,13 +32,15 @@ enum class MemoryType
    DEVICE_UMPIRE,  ///< Device memory; using Umpire
-   SIZE            ///< Number of host and device memory types
+   SIZE,           ///< Number of host and device memory types
+   PRESERVE        ///< Default parameter type; preserves existing behavior
```
Not sure that's the right place to put this enum - `KEEP` vs `PRESERVE` ?

tomstitt replied:
This was requested for cases like [this](https://github.com/mfem/mfem/blob/feature/tomstitt/temp-mem-type/linalg/sparsemat.hpp#L150-L151), to make it clear what the `mt` arg was doing in the default case

camierjs commented on general/mem_manager.hpp:
```diff
@@ -57,12 +59,14 @@ extern const char *MemoryTypeName[MemoryTypeSize];
+   HOST_64,     ///< Memory types: { HOST_64, HOST_DEBUG }
+   DEVICE,      /**< Memory types: { DEVICE, DEVICE_DEBUG, DEVICE_UMPIRE,
+                                     MANAGED } */
+   DEVICE_TEMP, ///< Memory types: { DEVICE_UMPIRE }
```
Why is there a correspondence between DEVICE_TEMP and DEVICE_UMPIRE ?

tomstitt replied:
Because the `DEVICE_UMPIRE` type implements a temporary memory allocator (`AllocTemp`)

camierjs commented on general/mem_manager.hpp:
```diff
@@ -137,16 +141,18 @@ class Memory
                                    Vector::UseDevice() */
-      ALIAS         = 1 << 7  ///< Pointer is an alias
+      ALIAS         = 1 << 7, ///< Pointer is an alias
+      USE_TEMPORARY = 1 << 8  ///< Temporary Device memory flag
```
We should unify `TEMP`, `TEMPORARY` to either `TEMP` or `POOL`

tomstitt replied:
I agree with that. For Umpire, `POOL` doesn't make sense because both spaces can be a pool. In our case we use one kind of Umpire Pool for the permanent device memory and another kind of pool for the temporary device memory. My vote is for `TEMP` but I'm okay with `POOL` if that's what others think is best.

The reason for the pool with permanent memory is there are still some small stack-allocated objects that own device memory and tagging all of them as temporary is a bit involved and would result in a more fragmented temporary pool, which we want to leave open for large temporary device allocations.

camierjs commented on general/mem_manager.hpp:
```diff
@@ -137,16 +141,18 @@ class Memory
-   int capacity; ///< Size of the allocated memory
-   MemoryType h_mt; ///< Host memory type
-   mutable unsigned flags; ///< Bit flags defined from the #FlagMask enum
+   T *h_ptr{nullptr};
```
MFEM's style is to initialize the variables in the constructor. We could change it, but not just for a few of them.

tomstitt replied:
Switched to `=` initializers !

artv3 commented:
@tomstitt I think we are missing some UMPIRE guards in cuda_umpire.cpp. Would we try to get travis passing?

tomstitt commented:
All of the CAMP package config has been removed!

tzanio commented on makefile:
```diff
@@ -332,8 +332,7 @@ MFEM_DEFINES = MFEM_VERSION MFEM_VERSION_STRING MFEM_GIT_STRING MFEM_USE_MPI\
  MFEM_USE_CEED MFEM_USE_RAJA MFEM_USE_UMPIRE MFEM_USE_SIMD MFEM_USE_ADIOS2\
- MFEM_USE_MKL_CPARDISO MFEM_USE_AMGX MFEM_USE_MUMPS MFEM_SOURCE_DIR\
- MFEM_INSTALL_DIR
+ MFEM_USE_MKL_CPARDISO MFEM_USE_AMGX MFEM_USE_MUMPS MFEM_SOURCE_DIR MFEM_INSTALL_DIR
```
Please revert

artv3 commented on mesh/mesh.cpp:
```diff
@@ -10751,16 +10751,19 @@ GeometricFactors::GeometricFactors(const Mesh *mesh, const IntegrationRule &ir,
    if (flags & GeometricFactors::COORDINATES)
    {
       X.SetSize(vdim*NQ*NE);
+      X.UseTemporary(use_temp_mem);
```
Minor comment, I was thinking this could be ``UseDevicePool`` or something of the sort since the following vector would also go away after it goes out of scope. 
```
{
  Vector X; 
}
````

artv3 replied:
Actually, use device pool may not make sense since you could have a permanent device pool, so temporary may make more sense.

v-dobrev commented:
I proposed some changes in #2050 -- we can discuss them there.

v-dobrev commented:
I tested (with `make test`) commit 716bde9 with CUDA and Umpire enabled and it worked fine.

v-dobrev commented:
LGTM

v-dobrev commented:
LGTM

tomstitt reviewed (commented):
LGTM

camierjs commented:
LGTM

camierjs reviewed (approved):
Thank you @tomstitt and @v-dobrev for pushing these new features!

v-dobrev commented:
Merged in `next` for testing...

tzanio commented:
Merging this, but there are diffs in the following runs with `master` that are not clear:

```
mpirun -np 4 ex3p -m ../data/star.mesh -pa -d cuda -no-vis
mpirun -np 4 ex3p -m ../data/star.mesh -pa -d raja-cuda -no-vis
mpirun -np 4 ex3p -m ../data/beam-hex.mesh -pa -d cuda -no-vis
```
